src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+                    raise CancelledError()
+                else:
+                    logging.error("Referencing failed: %s", ex)
+                    raise
+            except CancelledError:
+                logging.debug("Referencing cancelled.")
+                future._was_stopped = True
+                raise  # No fuss, pass it as-is
+            except Exception:
+                logging.exception("Referencing failure.")
+                raise
+            finally:
+                # If the state of the device is stable and all channels are valid, the referencing procedure
+                # succeeded. Typically, all channels become valid at the same time during the reference procedure.
+                # We still check all of them inidividually to be sure.
+                if self.IsStable():
+                    for ch, num in self.channels.items():
+                        self.referenced._value[ch] = self.IsValid(num)
+                self._updatePosition()
+                # We only notify after updating the position so that when a listener
+                # receives updates both values are already updated.
+                self.referenced.notify(self.referenced.value)
+
+                # The adjustment state should be "disabled" after automatic adjustment. If the referencing
+                # procedure was stopped prematurely, right after it was set to "manual", it might be in the
+                # wrong state --> make sure it's set to disabled here.
+                if self.GetAdjustmentState() != SA_SIDLL.SA_PS_ADJUSTMENT_STATE_DISABLED:
+                    self.SetProperty_i32(SA_SIDLL.SA_PS_AF_ADJUSTMENT_STATE_PROP,
+                                         SA_SIDLL.SA_PS_ADJUSTMENT_STATE_DISABLED)
+                    self.WaitForEvent(SA_SIDLL.SA_PS_ADJUSTMENT_STATE_DISABLED)  # state will be DISABLED when done
+                self.DisableEventNotification(SA_SIDLL.SA_PS_AF_ADJUSTMENT_PROGRESS_EVENT)
+
+    @isasync
+    def _validate(self):
+        """
+        Short "referencing" procedure. On startup, either referencing or validation needs to be performed.
+        Validation uses the referencing parameters stored in memory. It is in general much faster than
+        referencing. However, it is recommended to do the full referencing procedure on startup for
+        more accurate results.
+        """
+        f = self._createReferenceFuture()
+        self._executor.submitf(f, self._doValidation, f)
+        return f
+
+    def _doValidation(self, future):
+        """
+        Actually runs the validation code.
+        After the device was just turned on, it might take ~5 minutes, otherwise it is very fast.
+        future (Future): the future it handles
+        raise:
+            IOError: if validation failed due to hardware
+            CancelledError if was cancelled
+        """
+        with future._moving_lock:
+            try:
+                self.ValidateChannels()
```

Comment:
You could put this function inline here? It would look more similar to the referencing procedure.